src/platform/library/lib/memory.c

```diff
@@ -0,0 +1,234 @@
+static SHARED_DATA struct block_hdr rt_shared_block256[HEAP_RUNTIME_SHARED_COUNT256];
+static SHARED_DATA struct block_hdr rt_shared_block512[HEAP_RUNTIME_SHARED_COUNT512];
+static SHARED_DATA struct block_hdr rt_shared_block1024[HEAP_RUNTIME_SHARED_COUNT1024];
+
+/* Heap memory map for runtime shared */
+static SHARED_DATA struct block_map rt_shared_heap_map[] = {
+	BLOCK_DEF(64, HEAP_RUNTIME_SHARED_COUNT64, uncached_block_hdr(rt_shared_block64)),
+	BLOCK_DEF(128, HEAP_RUNTIME_SHARED_COUNT128, uncached_block_hdr(rt_shared_block128)),
+	BLOCK_DEF(256, HEAP_RUNTIME_SHARED_COUNT256, uncached_block_hdr(rt_shared_block256)),
+	BLOCK_DEF(512, HEAP_RUNTIME_SHARED_COUNT512, uncached_block_hdr(rt_shared_block512)),
+	BLOCK_DEF(1024, HEAP_RUNTIME_SHARED_COUNT1024, uncached_block_hdr(rt_shared_block1024)),
+};
+#endif
+
+/* Heap blocks for buffers */
+static SHARED_DATA struct block_hdr buf_block[HEAP_BUFFER_COUNT];
+static SHARED_DATA struct block_hdr lp_buf_block[HEAP_LP_BUFFER_COUNT];
+
+/* Heap memory map for buffers */
+static SHARED_DATA struct block_map buf_heap_map[] = {
+	BLOCK_DEF(HEAP_BUFFER_BLOCK_SIZE, HEAP_BUFFER_COUNT,
+		  uncached_block_hdr(buf_block)),
+};
+
+static SHARED_DATA struct block_map lp_buf_heap_map[] = {
+	BLOCK_DEF(HEAP_LP_BUFFER_BLOCK_SIZE, HEAP_LP_BUFFER_COUNT,
+		  uncached_block_hdr(lp_buf_block)),
+};
+
+static SHARED_DATA struct mm memmap;
+
+void platform_init_memmap(struct sof *sof)
+{
+	int i;
+
+	/* access memory map through uncached region */
+	sof->memory_map = cache_to_uncache(&memmap);
+
+	/* .system primary core initialization */
+	sof->memory_map->system[0].heap = (unsigned long)malloc(HEAP_SYSTEM_M_SIZE);
+	sof->memory_map->system[0].size = HEAP_SYSTEM_M_SIZE;
+	sof->memory_map->system[0].info.free = HEAP_SYSTEM_M_SIZE;
+	sof->memory_map->system[0].caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_EXT |
+		SOF_MEM_CAPS_CACHE;
+
+	/* .system_runtime primary core initialization */
+	sof->memory_map->system_runtime[0].blocks =
+		ARRAY_SIZE(sys_rt_heap_map[0]);
+	sof->memory_map->system_runtime[0].map =
+		uncached_block_map(sys_rt_heap_map[0]);
+	sof->memory_map->system_runtime[0].heap =
+		(unsigned long)malloc(HEAP_SYS_RUNTIME_M_SIZE);
+	sof->memory_map->system_runtime[0].size = HEAP_SYS_RUNTIME_M_SIZE;
+	sof->memory_map->system_runtime[0].info.free = HEAP_SYS_RUNTIME_M_SIZE;
+	sof->memory_map->system_runtime[0].caps = SOF_MEM_CAPS_RAM |
+		SOF_MEM_CAPS_EXT | SOF_MEM_CAPS_CACHE |
+		SOF_MEM_CAPS_DMA;
+
+	/* .system and .system_runtime secondary core initialization */
+	for (i = 1; i < CONFIG_CORE_COUNT; i++) {
+		/* .system init */
+		sof->memory_map->system[i].heap =
+				(unsigned long)malloc(HEAP_SYSTEM_S_SIZE) +
+			((i - 1) * SOF_CORE_S_SIZE);
+		sof->memory_map->system[i].size = HEAP_SYSTEM_S_SIZE;
+		sof->memory_map->system[i].info.free = HEAP_SYSTEM_S_SIZE;
+		sof->memory_map->system[i].caps = SOF_MEM_CAPS_RAM |
+			SOF_MEM_CAPS_EXT | SOF_MEM_CAPS_CACHE;
```

Comment:
`system[i]` initialization seems to be 80 or 90% identical to `system[0]`. It's impossible to spot and review the differences because of the big copy/paste/diverge. How about:
```C
   for (i = 1; i < CONFIG_CORE_COUNT; i++) {
           sof->memory_map->system[i] = sof->memory_map->system[0];

           /* Now initialize differences */
           sof->memory_map->system[i].heap =
				(unsigned long)malloc(HEAP_SYSTEM_S_SIZE) +
			((i - 1) * SOF_CORE_S_SIZE);
           ...
```

Reply:
ditto - this is tracking the cavs version.